### PR TITLE
Blog post: Push Button, Let's Encrypt

### PR DESCRIPTION
**Pull request by smashwilson**
Any feedback would be greatly appreciated! :bow:

### CODE REVIEW

File: _posts/2015-12-10-push-button-lets-encrypt.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
+---
+layout: post
+title: "Push Button, Let's Encrypt"
+date: 2015-12-10
+comments: true
+author: Ash Wilson <ash.wilson@rackspace.com>
+authorIsRacker: true
+published: true
+excerpt: >
+  An experiment to employ Let's Encrypt to serve any existing HTTP service over HTTPS with one docker command.
```
Comment (rgbkrk) [inline _posts/2015-12-10-push-button-lets-encrypt.md]: :+1:

File: _posts/2015-12-10-push-button-lets-encrypt.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+When [Let's Encrypt](https://letsencrypt.org/) went into public beta, I was curious to see how much farther I could lower the barrier to securing a service with strong TLS. The answer so far: [a single `docker run` command](https://github.com/smashwilson/lets-nginx).
+
+```bash
+docker run --detach \
+  --name lets-nginx \
+  --link backend:backend \
+  --env EMAIL=my@email.com \
+  --env DOMAIN=mydomain.io \
+  --env UPSTREAM=backend:8080 \
```
Comment (rgbkrk) [inline _posts/2015-12-10-push-button-lets-encrypt.md]: This is pretty slick.

File: _posts/2015-12-10-push-button-lets-encrypt.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+  --publish 80:80 \
+  --publish 443:443 \
+  smashwilson/lets-nginx
+```
+
+This container will invoke the [letsencrypt client](https://github.com/letsencrypt/letsencrypt) to issue a browser-trusted certificate valid for your domain. Then, it'll template a basic nginx configuration, a cronjob to re-issue the certificates monthly, and launch the crond and nginx daemons. Your service will have a green lock in browsers and an A+ score from [SSL Labs](https://www.ssllabs.com/ssltest/).
+
+![only the finest locally-source artisanal TLS]({% asset_path 2015-12-10-push-button-lets-encrypt/ssllabs.jpeg %})
```
Comment (rgbkrk) [inline _posts/2015-12-10-push-button-lets-encrypt.md]: always great to have beautiful alt text

File: _posts/2015-12-10-push-button-lets-encrypt.md [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+Some caveats before we begin: some encryption is a great improvement on no encryption, and TLS certificates signed by a trusted CA are more secure than self-signed ones, but trusting crypto in a container you download from strangers on the Internet does carry risks. I'm no security expert, but offhand I'd give this approach one and a half tin-foil hats out of five.
```
Comment (rgbkrk) [inline _posts/2015-12-10-push-button-lets-encrypt.md]: Wow this begs for a tin foil hat rating image.

File: _posts/2015-12-10-push-button-lets-encrypt.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+Furthermore, because this approach is automated, it uses the `--agree-tos` flag to accept the Let's Encrypt terms of service on your behalf. Be sure that you're okay with [the terms](https://letsencrypt.org/repository/) before you run the container.
+
+## Get Encrypted
+
+Before you get started, you will need:
+
+ 1. [A Carina cluster]({{ site.baseurl }}/docs/getting-started/getting-started-on-carina/) up, running and connected. It'll be easier if your cluster only has a single segment because the container's public IP will be predictable. Otherwise, you'll need to fiddle around with Swarm affinities.
+ 2. A domain name you own with a DNS A record pointing to your cluster's public IP.
+ 3. Any backend service that talks HTTP running in a Docker container on your cluster. I have a [Sinatra test service](https://github.com/smashwilson/minimal-sinatra) that's handy for iterating on infrastructure like this. Make a note of the container's name. It **does not** need to be listening on a public port; in fact, it's better if it isn't, because then you can funnel all traffic through the secure entrypoint instead.
```
Comment (rgbkrk) [inline _posts/2015-12-10-push-button-lets-encrypt.md]: woot

**Comment by rgbkrk** [comment]
For some workloads this basic nginx setup won't work properly, namely websockets and server sent events. I'm not real sure how they'd handle that without adapting lets-nginx themselves.

**Comment by smashwilson** [comment]
> For some workloads this basic nginx setup won't work properly, namely websockets and server sent events. I'm not real sure how they'd handle that without adapting lets-nginx themselves.

That's true. I should add that to the Limitations section, then.

**Comment by smashwilson** [comment]
@rgbkrk Tinfoil hat images added :wink:

screen shot 2015-12-10 at 3 43 05 pm

The markup will make @ktbartholomew cringe, but it'll do the job.

**Comment by rgbkrk** [comment]
YESSSSS

**Comment by rgbkrk** [comment]
I'm so using these assets in other posts. Pretty sure this means you need a javascript package for tinfoil hat declaration. Here's how you'd use the React component:

```jsx
<TinFoilHats
  rating={1.5}
/>
```

**Comment by rgbkrk** [comment]
Please don't hate me for saying that it looks like 1.3333333 hats.

**Comment by rgbkrk** [comment]
OMG when can we ship this?

Perhaps we should see if @kmsholcomb can have a look over.

**Comment by kmsholcomb** [comment]
I can start editing now, but will have to finish in the early evening.

**Comment by rgbkrk** [comment]
We might as well just submit this in the morning with new dates. I just got excited by the post and narrative.

**Comment by smashwilson** [comment]
> We might as well just submit this in the morning with new dates. I just got excited by the post and narrative.

Yep! Tomorrow morning is fine. Even though you promised Twitter I'd publish tonight. :wink:

**Comment by kmsholcomb** [comment]
Hi @smashwilson. The attached Word doc has my suggested copy edits and some comments. Please let me know if you have any questions. Thanks!

[Push Button Let's Encrypt_editsKH.docx](https://github.com/getcarina/getcarina.com/files/58963/Push.Button.Let.s.Encrypt_editsKH.docx)

**Comment by smashwilson** [comment]
Edits complete and post re-dated to today. Thanks :100:, @kmsholcomb!

**Comment by kmsholcomb** [comment]
You are welcome!

**Pull request merged**